pm/validator.go

```diff
@@ -0,0 +1,79 @@
+package pm
+
+import (
+	"fmt"
+	"math/big"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// Validator is an interface which describes an object capable
+// of validating tickets
+type Validator interface {
+	// ValidateTicket checks if a ticket is valid
+	ValidateTicket(ticket *Ticket, sig []byte, recipientRand *big.Int) error
+
+	// IsWinningTicket checks if a ticket won
+	// Note: This method does not check if a ticket is valid which is done using ValidateTicket
+	IsWinningTicket(ticket *Ticket, sig []byte, recipientRand *big.Int) bool
+}
+
+// validator is an implementation of the Validator interface
+// that relies on an implementation of the Broker interface to provide
+// a set of already used tickets
+type validator struct {
+	addr        ethcommon.Address
+	broker      Broker
+	sigVerifier SigVerifier
+}
+
+// NewValidator returns an instance of a validator
+func NewValidator(addr ethcommon.Address, broker Broker, sigVerifier SigVerifier) Validator {
+	return &validator{
+		addr:        addr,
+		broker:      broker,
+		sigVerifier: sigVerifier,
+	}
+}
+
+// ValidateTicket checks if a ticket is valid
+func (v *validator) ValidateTicket(ticket *Ticket, sig []byte, recipientRand *big.Int) error {
+	if ticket.Recipient != v.addr {
+		return fmt.Errorf("invalid ticket recipient")
+	}
+
+	if (ticket.Sender == ethcommon.Address{}) {
+		return fmt.Errorf("invalid ticket sender")
+	}
+
+	if crypto.Keccak256Hash(ethcommon.LeftPadBytes(recipientRand.Bytes(), uint256Size)) != ticket.RecipientRandHash {
+		return fmt.Errorf("invalid preimage provided for hash commitment recipientRandHash")
+	}
+
+	used, err := v.broker.IsUsedTicket(ticket)
```

Comment:
`IsUsedTicket` checks if the broker has already processed this particular winning ticket (i.e. query the smart contract). Here is a summary of an offline discussion about this:

The broker keeps track of used winning tickets in order to protect a sender from a malicious recipient that tries to replay a winning ticket over and over again. Since the validator is the recipient, we do not need to protect against this scenario. However, we do need to protect against the scenario in which a sender purposefully sends used winning tickets to the recipient which results in the recipient being unable to redeem them. 

If we assume that the recipient also:

- Handles replay prevention for invalid already winning `recipientRand` values
- Uses an HMAC based on a secret generated on start up and the sender's address to validate `recipientRand` values

then the recipient does not need to call `IsUsedTicket` when validating a ticket.

The recipient will not accept tickets with an old winning `recipientRand` which means no old winning tickets created by the current sender will be accepted. Sender A can try to use winning tickets sent by sender B, but these tickets will not pass the HMAC validation step since the tickets include a different sender. If the recipient restarts, thereby clearing its current set of invalid already winning `recipientRand` values, it will generate a new secret so if a sender tries to replay an old `recipientRand` value, it will fail the HMAC validation which uses the recipient's new secret as the key.